Add "efs-dir" provsioning mode

mpatlasov opened the pull request:
This PR was inspired by [Jonathan Rainer](https://github.com/jonathanrainer)'s [PR](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/732), and [Fabio Bertinatto](https://github.com/bertinatto)'s [PR](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1408).

**Is this a bug fix or adding new feature?**

New feature, requested in https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/538 and https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/517 and possibly other issues, this comes up a lot as a feature people would like.

**What is this PR about? / Why do we need it?**

Since its creation the driver has supported Access Point Provisioning as its main means of dynamic provisioning. However this is problematic for a few reasons, it can cause issues with deleting access-points (as reported [here](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/517)) and it requires various AWS IAM Permissions that can be complicated to sort out and manage. Customers want a way to set their own permissions from the top level directory, but the mode doesn't allow to do that.

This PR allows a new provisioning mode called efs-dir so that instead of creating EFS access points, directories are created instead. This is achieved by creating a new Interface called Provisioner that is implemented by an AccessPointProvisioner (the original method) and a DirectoryProvisioner (the new method) this also allows in future for different kinds of provisioning to occur, maybe FileSystemProvisioning for example.

**What testing is done?** 
New code is covered by unit-tests, they all pass when I run "go test" locally. New e2e tests are added covering create/delete operation in "efs-dir" mode. They also pass when I run them against OpenShift (OCP) cluster with new driver installed.

fixes https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/538

mpatlasov committed:
Simplify controller_test.go

Use `buildDriver()` instead of `&Driver{...}`

mpatlasov commented:
/test pull-aws-efs-csi-driver-unit

seanzatzdev-amazon commented:
Please rebase

mpatlasov commented:
Update: last force-push rebased PR (only).

mpatlasov commented:
Update: last force-push fixed e2e tests for "efs-dir".

mpatlasov commented:
/test pull-aws-efs-csi-driver-unit

dankova22 reviewed (commented):
@mpatlasov  Thanks for the PR, we are considering supporting this feature. If could you address these comments and rebase, that would be great

mpatlasov committed:
Update charts, examples and documentation

mpatlasov committed:
Add e2e tests for "efs-dir" provisioning mode.

mpatlasov commented:
Hi @dankova22 , thank you for reviewing this PR, highly appreciated! I've just rebased it, and not addressed your comments yet. Will do it soon.

mpatlasov commented:
> Hi @mpatlasov Im interested to hear more about one of your concerns with current access point provisioning:
> 
> > it can causes issues with user permissions around deleting provisioned directories
> 
> Is this referring to issues you have faced deleting access point root dirs outside of k8 or when deleting volumes with `deleteAccessPointRootDir`?

This phrase in PR's description is copy/pasta from [Jonathan Rainer](https://github.com/jonathanrainer)'s [PR](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/732) description. I thought it referred the issue https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/517:

> What happened?
> Access Points are not being deleted all PVCs in the namespace are deleted
> What you expected to happen?
> Access Points needs ro deleted when corresponding PVC is deleted

I have to remove "user permissions around deleting provisioned directories" from the description, sorry for that confusion.

mpatlasov commented:
/hold

mpatlasov commented:
/test pull-aws-efs-csi-driver-e2e

mpatlasov committed:
Avoid redundant call `DescribeMountTargets` in dir-mode Provision()

mpatlasov committed:
In dir-mode Delete(), remove temporary directory only if empty

mpatlasov committed:
Update README to highlight storage class parameters usable only in "efs-ap" mode.

mpatlasov commented:
/unhold

mpatlasov commented:
Hi @dankova22 , all comments addressed, PTAL please!

dankova22 commented on docs/README.md:
```diff
@@ -25,22 +25,27 @@ The following CSI interfaces are implemented:
+| reuseAccessPoint      |                | false           | true     | When set to true, it creates the Access Point client-token from the provided PVC name. So that the AccessPoint can be replicated from a different cluster if same PVC name and storageclass configuration are used. Used only for provisioningMode "efs-ap".                                                                            ...
 
 **Note**
+* We suggest the following settings for each provisioning mode:
+  * `efs-ap` -> `directoryPerms: 770`, with `gidRangeStart` and `gidRangeEnd` set to sensible values
+  * `efs-dir` -> `directoryPerms: 777` and omit `gidRangeStart` and `gidRangeEnd`
+  * This ensures that important system utilities will still be to access to the directories created
+  * All parameters are available in all modes but please think very carefully about what you're doing otherwise directories could be created that no-one can read or write to.
```
I dont think we should recommend setting permissive 777 permissions (although this is the easiest to manage and users may choose to use this).

I think the recommendation should be for users to either:

1. Set a static gid in the storage class, and add this is as a supplementary group for all pods that will use this storage class.

2. Alternatively, users could use fsGroup to have kubernetes handle ownership and pod access to provisioned dirs. Key points would be:

- Need to set [fsGroup in securityContext for pod ](https://kubernetes-csi.github.io/docs/support-fsgroup.html)
- Need to ensure [fsGroupPolicy in CSI Driver](https://kubernetes-csi.github.io/docs/support-fsgroup.html) spec is set to mode that supports this (File or ReadWriteOnceWithFSType). This will allow kubernetes to change the ownership of the provisioned dir.
- For RWX cases if doing this, users should be extra careful to:
	  - Ensure all pods using same volume run with the same fsGroup, otherwise, one of their pods may be locked out as ownership will change.
	  - In Pod security context, Set [fsGroupChangePolicy as "OnRootMismatch"](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods) to avoid expensive recursive chowns of provisioned dir. 

We should have some documentation about this in the efs-dir example README, that we can reference here.

dankova22 commented on pkg/driver/provisioner_dir.go:
```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not unmount %q: %v", target, err)
+	}
+	err = d.osClient.RemoveAll(target)
```
We should use Remove here too instead of RemoveAll

consider the case where thread A mounts the root of the filesystem at /var/lib/csi/pv/1234

Thread B executes removeAll /var/lib/csi/pv/1234. We would delete the whole fs. 

The chance of uuid collision is pretty much zero, but better to use Remove to limit blast radius in case of a bug as the target should be empty at this point anyway

dankova22 commented on pkg/driver/provisioner_dir.go:
```diff
@@ -0,0 +1,229 @@
+		// If that fails then track the error but don't do anything else
+		if unmountErr != nil {
+			klog.V(5).Infof("Unmount failed at '%s'", target)
+			e = status.Errorf(codes.Internal, "Could not unmount %q: %v", target, err)
```
In this case we never log or return the unmountErr. I think we prob want to do 

e = status.Errorf(codes.Internal, "Could not unmount %q: %v", target, unmountErr)

We should also make sure we log the errors from Mount and RemoveAll, so that we would have visibility in the case when both cleanup and core functionality fails.

dankova22 commented on test/e2e/e2e.go:
```diff
@@ -416,6 +418,93 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
 			}
 		})
 
+		createProvisionedDirectory := func(f *framework.Framework, basePath string, pvcName string) (*v1.PersistentVolumeClaim, *storagev1.StorageClass) {
```
We need to add e2e test coverage of a actually writing to the fs. I think a test similar to "should successfully perform dynamic provisioning" for "efs-dir" is sufficient.

dankova22 reviewed (commented):
Thank you for the continued effort on this and addressing the previous comments.

I left a couple more thoughts after reviewing again. Please let me know if you are able to get to that or I will try to find some time to address myself.

mpatlasov commented:
Thank you for reviewing again, @dankova22 ! Can you please compose a patch for [docs/README.md](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1497#discussion_r1938646754), and I'll work on addressing other comments meanwhile.

mpatlasov committed:
In dir-mode Provision(), remove temporary directory only if empty

mpatlasov committed:
Improve logging in defer func of Delete() for "efs-dir" mode

Two issues fixed:
 * If Unmount() failed, return error message with unmountErr
 * If Delete() failed before defer func, and defer func failed, return error message with both errors

mpatlasov committed:
Add e2e test writing to the fs in "dir-mode" provsioning case.

mpatlasov commented:
Hi @dankova22 , all comments (excepting one about  [docs/README.md](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1497#discussion_r1938646754)) are addressed. Let me know if you can fix docs/README.md.

dankova22 commented:
Thanks again! Here is a patch for the documentation update:
[update-readme-directory-provisioning.patch](https://github.com/user-attachments/files/18916700/update-readme-directory-provisioning.patch)

mpatlasov committed:
Add new section (Directory Provisioning Mode) to README

Thanks Dan for [the patch](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1497#issuecomment-2675767085).

mpatlasov commented:
Hi @dankova22 , thanks for the doc patch, I've applied it now (sorry for delay - I was afk for last week).

mpatlasov committed:
Remove deprecated framework.ExpectEqual from e2e.go

mskanth972 commented:
/lgtm

mpatlasov committed:
Introduce Provisioner interface

Controller's CreateVolume and DeleteVolume now call method of Provisioner interface. All heavy-lifting logic of create/delete specific for access-point provsioning is hidden now in AccessPointProvisioner struct.

mpatlasov committed:
Add directory provisioning mode

Implement `DirectoryProvisioner` struct and its `Provision/Delete` methods. Add `delete-provisioned-dir` command-line option to control DeleteVolume behavior in efs-dir mode.

mpatlasov committed:
Add unit-tests for directory and access-point provisioning modes

The patch adds new unit-tests for `provisioner_ap.go` and `provisioner_dir.go`, and also adds a few unit-tests for `controller.go`.

mpatlasov committed:
Add support for static uid/gid from storage class for dir-mode provisioning

mpatlasov commented:
Update: last force-push rebased PR (only).

mpatlasov commented:
`pull-aws-efs-csi-driver-external-test-eks` and `pull-aws-efs-csi-driver-e2e` are perma-failing:
```
ERROR: failed to solve: process "/bin/sh -c clean_install amazon-efs-utils true &&     clean_install crypto-policies true &&     install_binary         /usr/bin/cat         /usr/bin/cd         /usr/bin/df         /usr/bin/env         /usr/bin/find         /usr/bin/grep         /usr/bin/ln         /usr/bin/ls         /usr/bin/mount         /usr/bin/umount         /sbin/mount.nfs4         /usr/bin/openssl         /usr/bin/sed         /usr/bin/stat         /usr/bin/stunnel         /usr/sbin/tcpdump         /usr/bin/which &&     cleanup \"efs-csi\"" did not complete successfully: exit code: 1
make[1]: *** [Makefile:86: .image-30814-linux-amd64-amazon] Error 1
make[1]: Leaving directory '/home/prow/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver'
make: *** [Makefile:119: test-e2e] Error 2
```
This is not relevant to PR.

mpatlasov commented:
/remove-lifecycle stale
